tedana/utils/utils.py

```diff
@@ -1,198 +1,452 @@
     ----------
-    data : array_like
-        Masked array of shape (nx*ny*nz[, Ne[, nt]])
-    mask : array_like
-        Boolean array of shape (nx, ny, nz)
+    data : (X x Y x Z x E x T) array_like
+        Multi-echo data array
 
     Returns
     -------
-    ndarray
-        Array of shape (nx, ny, nz[, Ne[, nt]])
+    fdata : (X x Y x M x T) np.ndarray
+        Z-concatenated multi-echo data array, where M is Z * number of echos
     """
 
-    M = (mask != 0).ravel()
-    Nm = M.sum()
+    if data.ndim < 4:
+        raise ValueError('Input data must have at least four dimensions; '
+                         'provided data has only {0}'.format(data.ndim))
 
-    nx, ny, nz = mask.shape
+    (nx, ny), nz = data.shape[:2], np.prod(data.shape[2:4])
+    return data.reshape(nx, ny, nz, -1)
 
-    if len(data.shape) > 1:
-        nt = data.shape[1]
-    else:
-        nt = 1
 
-    out = np.zeros((nx * ny * nz, nt), dtype=data.dtype)
-    out[M, :] = np.reshape(data, (Nm, nt))
+def fmask(data, mask=None):
+    """
+    Masks `data` with non-zero entries of `mask`
 
-    return np.squeeze(np.reshape(out, (nx, ny, nz, nt)))
+    Parameters
+    ----------
+    data : (X x Y x Z [x E [x T]) array_like or img_like object
+        Data array or data file to be masked
+    mask : (X x Y x Z) array_like or img_like object
+        Boolean array or mask file
+
+    Returns
+    -------
+    fdata : (S x E x T) np.ndarray
+        Masked `data`, where `S` is samples, `E` is echoes, and `T` is time
+    """
+
+    if mask is not None and not type(data) == type(mask):
+        raise TypeError('Provided `data` and `mask` must be of same type.')
+
+    if isinstance(data, str):
+        root, ext, addext = splitext_addext(data)
+        if ext == '.gii':
+            # mask need not apply for gii files
+            fdata = np.column_stack([f.data for f in nib.load(data).darrays])
+        else:
+            # use nilearn for other files
+            data = check_niimg(data)
+            if mask is not None:
+                # TODO: check that this uses same order to flatten
+                fdata = nimask.apply_mask(data, mask).T
+            else:
+                fdata = data.get_data().reshape((-1,) + data.shape[3:])
+    elif isinstance(data, np.ndarray):
+        # flatten data over first three dimensions and apply mask
+        fdata = data.reshape((-1,) + data.shape[3:])
+        if mask is not None:
+            fdata = fdata[mask.flatten() > 0]
+
+    return fdata.squeeze()
+
+
+def unmask(data, mask):
+    """
+    Unmasks `data` using non-zero entries of `mask`
+
+    Parameters
+    ----------
+    data : (M x E x T) array_like
+        Masked array, where `M` is the number of samples
+    mask : (S,) array_like
+        Boolean array of `S` samples that was used to mask `data`
+
+    Returns
+    -------
+    out : (S x E x T) np.ndarray
+        Unmasked `data` array
+    """
+
+    out = np.zeros((mask.shape + data.shape[1:]))
```

Comment:
Is there a reason this is not `out = np.zeros(mask.shape + data.shape[1:])` ?

Reply:
Nope! Good catch -- will fix in the next commit.